# Exploring directories using shell

Multiple uses of / is as good as one
The root folder / is its own parent 

clear  used to clear the screen or use ctrl+L

exit or ctrl+D to come out of shell

*****************************************************
HELP

ls --help
summary of how to use the command

man ls
manual page for the command, which provides detailed documentation.

**************************************************


*****************************************************************
Filesystem Hierarchy Standard followed by all linux systems
FHS 3.0 released on June 03, 2015
Available at
https://refspecs.linuxfoundation.org/fhs.shtml

*****************************************************************

In [3]:
%%bash
pwd
# Present working directory
# /mnt/c/ indicates that the path is being accessed through WSL (Windows Subsystem for Linux).
# In WSL, Windows drives (e.g., C:, D:) are mounted under the /mnt directory. Here, /mnt/c/ corresponds to the C: drive in Windows.

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell


In [4]:
%%bash
ls
#List files in current folder

1500x500.jpeg
Linux_shell_setup_windows.ipynb
Linux_shell_w1.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
level1
random_toss.py


In [5]:
%%bash
ls -m
#This command lists the files in a horizontal format, with each filename separated by a comma and space.

1500x500.jpeg, Linux_shell_setup_windows.ipynb, Linux_shell_w1.ipynb,
SC_Week0_Unit0.ipynb, SC_Week0_Unit1.ipynb, SC_Week0_Unit2.ipynb,
SC_Week0_Unit4.ipynb, SC_Week0_Unit5.ipynb, level1, random_toss.py


In [15]:
%%bash
ls | tr '\n' ' '
#tr '\n' ' ': Translates newlines (\n) to spaces (' '), effectively putting all filenames in one line.

1500x500.jpeg Linux_shell_setup_windows.ipynb Linux_shell_w1.ipynb SC_Week0_Unit0.ipynb SC_Week0_Unit1.ipynb SC_Week0_Unit2.ipynb SC_Week0_Unit4.ipynb SC_Week0_Unit5.ipynb level1 random_toss.py 

In [21]:
%%bash
ls -am
# -a option gives hidden files/directories also

., .., .hidden_directory, .hidden_file.txt, 1500x500.jpeg,
Linux_shell_setup_windows.ipynb, Linux_shell_w1.ipynb, SC_Week0_Unit0.ipynb,
SC_Week0_Unit1.ipynb, SC_Week0_Unit2.ipynb, SC_Week0_Unit4.ipynb,
SC_Week0_Unit5.ipynb, level1, random_toss.py


In [25]:
%%bash
ls -l
# -l option gives long format

total 584
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 132170 Oct  8 02:55 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  6 18:19 level1
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py


In [87]:
%%bash
cd /    #root directory  (/) is the highest level in the filesystem hierarchy 
pwd
ls -am

#When you list all items in a directory using ls -a, you will see . and .. and hidden ones also starting with .
#   . represents the current directory.
#  .. always represents the parent directory, even if it’s a reference to the same directory (as is the case in the root directory).

/
., .., bin, bin.usr-is-merged, boot, dev, etc, home, init, lib,
lib.usr-is-merged, lib64, lost+found, media, mnt, opt, proc, root, run, sbin,
sbin.usr-is-merged, snap, srv, sys, tmp, usr, var


In [75]:
%%bash

# 1. **Starting in the Home Directory**:
# Print the current directory using `pwd`
pwd

# 2. **Change to the Current Directory (`cd .`)**:
# `cd .` changes to the current directory (no effect).
cd .
pwd

# 3. **Change to the Parent Directory (`cd ..`)**:
# `cd ..` moves up one directory level.
# Create a sample directory structure first for demonstration.
mkdir -p demo/level1/level2
cd demo/level1/level2
pwd               # Print current directory path

cd ..
pwd                # Move up to the parent directory (from level2 to level1)

cd ..
pwd                # Move up again to the parent directory (from level1 to demo)

cd ..              # come back to original directory and delete the demo directory
pwd
rm -r demo



/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/demo/level1/level2
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/demo/level1
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/demo
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell


In [74]:
%%bash

# 4. **Change to the Home Directory (`cd` or `cd ~`)**:
# `cd` or `cd ~` moves to the home directory.
cd
pwd                # Should print the home directory path

cd ~
pwd                # Again, should print the home directory path

# 5. **Using Environment Variable (`cd $HOME`)**:
# `cd $HOME` also navigates to the home directory.
cd $HOME
pwd                # Should print the home directory path again



# 6. **Absolute Path**:
# Change directory using the absolute path (full path from the root directory).
# Absolute paths start with `/`. Example: /home/username/demo
cd /            # Change to the root directory using absolute path
pwd

mkdir -p $HOME/demo/level1/level2
# Go to the previous demo directory using an absolute path
cd $HOME/demo
pwd                # Should print /home/<username>/demo (full path)
rm -r $HOME/demo


/home/kavis
/home/kavis
/home/kavis
/
/home/kavis/demo


In [73]:
%%bash
# 7. **Relative Path**:
# Change directory using relative paths, which do not start with `/`.
# Relative paths are based on the current working directory.
pwd
mkdir -p demo/leve1/level2
# Go to level1 directory from demo using relative path
cd demo
cd level1
pwd                # Should print the path ending in /demo/level1

# Go back to demo using relative path
cd ..
pwd                # Should print the path ending in /demo

cd ..
rm -r demo

# 8. **Navigate to Root Directory (`cd /`)**:
# Change to the root directory using `cd /`.
cd /
pwd                # Should print the root directory (`/`)
ls -m


/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/demo/level1
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/demo
/
bin, bin.usr-is-merged, boot, dev, etc, home, init, lib, lib.usr-is-merged,
lib64, lost+found, media, mnt, opt, proc, root, run, sbin, sbin.usr-is-merged,
snap, srv, sys, tmp, usr, var
/tmp/explore
/tmp/explore/dirA
/tmp/explore/dirA/dirB
/tmp/explore/dirA
/home/kavis
/tmp/explore/dirA
/tmp/explore/dirA


In [84]:
%%bash
# 9. **Create and Navigate Between Directories**:
# Create a new directory structure in root directory for further exploration
pwd
mkdir -p /tmp/explore/dirA/dirB/dirC
cd /tmp/explore
pwd                # Current path is /tmp/explore

# Go into dirA using a relative path
cd dirA
pwd                # Current path is /tmp/explore/dirA

# Go into dirB using a relative path
cd dirB
pwd                # Current path is /tmp/explore/dirA/dirB

# Use `cd ..` to go up one level
cd ..
pwd                # Current path is /tmp/explore/dirA

# Go to the home directory using `cd ~`
cd ~
pwd                # Should print the home directory path again

# 10. **Return to the Previous Directory (`cd -`)**:
# `cd -` returns to the previous directory.
cd -
pwd                # Should return you back to /tmp/explore/dirA

cd ../../..
pwd                # Should return you back to root directory

# Clean up: Remove the created directories for demonstration
rm -r /tmp/explore



/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/tmp/explore
/tmp/explore/dirA
/tmp/explore/dirA/dirB
/tmp/explore/dirA
/home/kavis
/tmp/explore/dirA
/tmp/explore/dirA
/


# ls command


In [141]:
%%bash

cd
pwd

# Create a test directory and navigate into it
mkdir -p test_dir/level1
cd test_dir
pwd

# Create some files and subdirectories
touch file1 file2 file3
mkdir subdir1 subdir2
ls # Default listing of files and directories in the current directory

# List in long format (-l) showing permissions, owner, group, size, and timestamp
ls -l

# List with human-readable file sizes (-lh) to display sizes in KB, MB, etc.
ls -lh

# Show hidden files as well (-a), includes files starting with '.'
ls -a

# Long format with hidden files (-la), useful to see permissions of hidden files
ls -la

# List in long format with hidden files and include inode number (-lia)
ls -lia

# Show the inode numbers (-i) of files and directories in the current directory
ls -i

# List in long format of only directories, showing details of directories themselves (-ld)
ls -ld */

# Show inode numbers along with hidden files in a directory listing (-ia)
ls -ia

# Show inode numbers along with detailed view for hidden files (-lia)
ls -lia

# Show file sizes in blocks (-s)
ls -s

# Recursive listing, show contents of all directories and subdirectories (-R)
ls -R

# Recursive listing in long format (-lR)
ls -lR

# Show files sorted by size (-lS) to get largest files first
ls -lS

# List files sorted by modification time, newest first (-lt)
ls -lt

# List files in reverse order by modification time (-ltr)
ls -ltr

# Combine -i and -l to show inode numbers along with detailed view (-il)
ls -il

# Combine -i, -a, and -l to show inode numbers, hidden files, and detailed view (-ial)
ls -ial

# Using a directory as an argument, list contents of "level1" directory
ls level1

# Navigate into the "level1" directory and list contents
cd level1
pwd
ls -l

# Go back to parent directory and list contents recursively with details (-lR)
cd ..
ls -lR

# Use -d to show information only about the directory itself, not its contents
ls -ld level1

# List all files including hidden ones and inodes, but just at the top-level directory (-iAd)
ls -iAd *

# List files with detailed view and human-readable sizes, sort by size (-lhS)
ls -lhS

# List directories themselves in detailed view with inode numbers (-ild)
ls -ild */

# Clean up: remove the created test directory
cd ..
rm -r test_dir


/home/kavis
/home/kavis/test_dir
file1
file2
file3
level1
subdir1
subdir2
total 12
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:16 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:16 file2
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:16 file3
drwxr-xr-x 2 kavis kavis 4096 Oct  8 10:16 level1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 10:16 subdir1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 10:16 subdir2
total 12K
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:16 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:16 file2
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:16 file3
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 10:16 level1
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 10:16 subdir1
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 10:16 subdir2
.
..
file1
file2
file3
level1
subdir1
subdir2
total 20
drwxr-xr-x  5 kavis kavis 4096 Oct  8 10:16 .
drwxr-x--- 10 kavis kavis 4096 Oct  8 10:16 ..
-rw-r--r--  1 kavis kavis    0 Oct  8 10:16 file1
-rw-r--r--  1 kavis kavis    0 Oct  8 10:16 file2
-rw-r--r--  1 kavis kavis    0 Oct  8 10:16 file3
drwxr-xr-x 

In [ ]:
%%bash

cd
pwd

# Create a test directory structure with nested subdirectories and files
mkdir -p test_dir/level1/level2
cd test_dir
touch file1 file2 level1/file3 level1/level2/file4
mkdir subdir1 subdir2

# Default listing of the current directory
ls

# List contents of a specific directory using it as an argument
ls level1

# List contents of multiple directories (subdir1, level1) using directory names as arguments
ls subdir1 level1

# List contents of multiple directories in long format (-l)
ls -l subdir1 level1

# List the contents of level1 recursively (-R), showing all files in subdirectories
ls -R level1

# List detailed view of the level1 directory itself without showing its contents (-ld)
ls -ld level1

# Using -ld to show information about the directory "level1/level2" without listing its contents
ls -ld level1/level2

# List with detailed view and human-readable sizes (-lh) for multiple directories
ls -lh level1 subdir1

# Recursive listing for multiple directories (-lR)
ls -lR level1 subdir2

# Combine options to show all directories and subdirectories recursively with human-readable sizes (-alhR)
ls -alhR level1 subdir1 subdir2

# Use a combination of -i (inode numbers) and directory arguments
ls -i level1 level1/level2

# List inode numbers and long format with human-readable sizes for multiple directories (-ilh)
ls -ilh level1 subdir1 subdir2

# Clean up: remove the created test directory
cd ..
rm -r test_dir


/home/kavis
file1
file2
level1
subdir1
subdir2
file3
level2
level1:
file3
level2

subdir1:
level1:
total 4
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:47 level2

subdir1:
total 0
level1:
file3
level2

level1/level2:
file4
drwxr-xr-x 3 kavis kavis 4096 Oct  8 02:47 level1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:47 level1/level2
level1:
total 4.0K
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:47 level2

subdir1:
total 0
level1:
total 4
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:47 level2

level1/level2:
total 0
-rw-r--r-- 1 kavis kavis 0 Oct  8 02:47 file4

subdir2:
total 0
level1:
total 12K
drwxr-xr-x 3 kavis kavis 4.0K Oct  8 02:47 .
drwxr-xr-x 5 kavis kavis 4.0K Oct  8 02:47 ..
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:47 level2

level1/level2:
total 8.0K
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:47 .
drwxr-x

In [152]:
%%bash

cd
pwd

# Create a sample directory structure
mkdir -p sample_dir/subdir
cd sample_dir
touch file1 file2 .hidden_file subdir/file3

# Short form of options:
ls -a          # Show all files, including hidden files
ls -l          # Show long format listing
ls -la         # Combine -l and -a to show hidden files in long format
ls -lh         # Long format with human-readable file sizes
ls -R          # Recursively list all directories and their contents
ls -d */       # Show directory names only
ls -i          # Show inode numbers
ls -r          # Reverse the order of the listing
ls -S          # Sort by file size
ls -t          # Sort by modification time

# Long form of options:
ls --all           # Equivalent to -a, show all files including hidden ones
ls --long          # Equivalent to -l, show long format listing
ls --all --long    # Equivalent to -la, show hidden files in long format
ls --human-readable --long  # Long format with human-readable sizes
ls --recursive     # Equivalent to -R, show directories and their contents recursively
ls --directory */  # Equivalent to -d, show directory names only
ls --inode         # Equivalent to -i, show inode numbers
ls --reverse       # Equivalent to -r, show files in reverse order
ls --sort=size     # Equivalent to -S, sort by file size
ls --time          # Equivalent to -t, sort by modification time

# Clean up: remove the created test directory
cd ..
rm -r sample_dir


/home/kavis
.
..
.hidden_file
file1
file2
subdir
total 4
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:31 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:31 file2
drwxr-xr-x 2 kavis kavis 4096 Oct  8 10:31 subdir
total 12
drwxr-xr-x  3 kavis kavis 4096 Oct  8 10:31 .
drwxr-x--- 11 kavis kavis 4096 Oct  8 10:31 ..
-rw-r--r--  1 kavis kavis    0 Oct  8 10:31 .hidden_file
-rw-r--r--  1 kavis kavis    0 Oct  8 10:31 file1
-rw-r--r--  1 kavis kavis    0 Oct  8 10:31 file2
drwxr-xr-x  2 kavis kavis 4096 Oct  8 10:31 subdir
total 4.0K
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:31 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 10:31 file2
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 10:31 subdir
.:
file1
file2
subdir

./subdir:
file3
subdir/
83918 file1
83919 file2
83917 subdir
subdir
file2
file1
subdir
file1
file2
file1
file2
subdir
.
..
.hidden_file
file1
file2
subdir
.:
file1
file2
subdir

./subdir:
file3
subdir/
83918 file1
83919 file2
83917 subdir
subdir
file2
file1
subdir
file1
file2


ls: unrecognized option '--long'
Try 'ls --help' for more information.
ls: unrecognized option '--long'
Try 'ls --help' for more information.
ls: unrecognized option '--long'
Try 'ls --help' for more information.
ls: option '--time' requires an argument
Try 'ls --help' for more information.


Key Points to Remember:
Short Form: Typically a single character preceded by a hyphen (-), e.g., -l, -a, -R.

Long Form: A descriptive word preceded by two hyphens (--), e.g., --all, --recursive, --long.

Short and long forms can be used interchangeably. For example, -l is equivalent to --long.

Multiple short options can be combined (e.g., ls -laR), but long options must be written separately (e.g., ls --long --all --recursive).

Use man ls or ls --help to see all available short and long forms for ls.

Short Form vs. Long Form of ls Options

Short Form	Long Form	Description
-a	          --all	    Show all files, including hidden files (those starting with .).

-A	       --almost-all	Show all files except . (current directory) and .. (parent).

-d	     --directory	List directory names only, not their contents.

-h	--human-readable	Show sizes in human-readable format (KB, MB, etc.).

-l	          --long	Use a long listing format to show detailed information.
-R	     --recursive	List subdirectories recursively.

-S	     --sort=size	Sort files by size (largest first).

-t	          --time	Sort by modification time (newest first).

-r	--reverse	Reverse the order of sorting.

-i	--inode	Show inode numbers.

-s	--size	Display file sizes in blocks.

-u	--access-time	Sort by last access time instead of modification time.

-c	--time=ctime	Sort by change time (status change time).

-G	--no-group	Do not show group information in the long format.

-X	--sort=extension	Sort alphabetically by file extension.

-1	--format=single-column	List one file per line.

This sequence includes a broad range of options for ls:

-l: Long format.
-h: Human-readable sizes.
-a: Include hidden files.
-i: Show inode numbers.
-d: Show directory information only.
-R: Recursive listing.
-S: Sort by size.
-t: Sort by modification time.
-r: Reverse the order.
-ia: Include inode numbers and hidden files.

Explanation of ls with Directories as Arguments:


ls <directory_name>: Lists the contents of the specified directory, not the current working directory.
Example: ls level1 lists files inside level1 directory.


ls <dir1> <dir2>: Lists contents of multiple directories.
Example: ls level1 subdir1 lists files in both level1 and subdir1.


ls -ld <directory>: Lists information about the directory itself, not its contents.
Example: ls -ld level1 shows details (like permissions, owner, etc.) of the directory level1 itself.


Recursive Listing (ls -R <directory>): Displays all files and subdirectories within the specified directory recursively.
Example: ls -R level1 shows all files in level1 and its subdirectories.


Combine Options with Directory Arguments:
ls -alhR level1 subdir1: Shows all hidden files in a long format, human-readable sizes, and recursively lists contents of level1 and subdir1.

In [140]:
%%bash
pwd
cd /mnt/c
ls -l
clear

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell


ls: DumpStack.log: Permission denied
ls: cannot access 'DumpStack.log.tmp': Permission denied
ls: cannot access 'hiberfil.sys': Permission denied
ls: cannot access 'pagefile.sys': Permission denied
ls: PerfLogs: Permission denied
ls: Recovery: Permission denied
ls: cannot access 'swapfile.sys': Permission denied
ls: 'System Volume Information': Permission denied


total 16
drwxrwxrwx 1 kavis kavis  4096 Oct 20  2022 $Recycle.Bin
-r-xr-xr-x 1 kavis kavis    40 Oct 20  2022 2CA0AE717CAA
lrwxrwxrwx 1 kavis kavis    12 Oct 10  2022 Documents and Settings -> /mnt/c/Users
drwxrwxrwx 1 kavis kavis  4096 Jul 30  2022 Drivers
---------- 1 kavis kavis 12288 Jul 12 03:44 DumpStack.log
-????????? ? ?     ?         ?            ? DumpStack.log.tmp
drwxrwxrwx 1 kavis kavis  4096 May  4 08:13 OneDriveTemp
d--x--x--x 1 kavis kavis  4096 May  7  2022 PerfLogs
dr-xr-xr-x 1 kavis kavis  4096 Oct  6 03:43 Program Files
dr-xr-xr-x 1 kavis kavis  4096 Jun 11 03:17 Program Files (x86)
drwxrwxrwx 1 kavis kavis  4096 Sep 19 16:32 ProgramData
d--x--x--x 1 kavis kavis  4096 Aug 10  2023 Recovery
-r-xr-xr-x 1 kavis kavis  1505 Dec 19  2023 Setup.log
d--x--x--x 1 kavis kavis  4096 Oct  7 17:23 System Volume Information
drwxrwxrwx 1 kavis kavis  4096 Dec  3  2023 Tenorshare
dr-xr-xr-x 1 kavis kavis  4096 Jul 10 09:50 Users
dr-xr-xr-x 1 kavis kavis  4096 Oct  5 11:03 Windows



#### Structure of Each Line given by ls -l:
Structure will change slightly depending on VS code is run in wsl or not

1. **Permissions**: (`drwxr-xr-x`)
   - This shows the file permissions for the owner, group, and others.
   - first letter -  `d` indicates a **directory** (if it were a file, it would be `-`).
                     'l' indicates symbolic link (shortcut)
                     '-' indicates file 
                     'c' character device (terminal)
                     'b' block devices (hard discs)
   - next three for owner - `rwx` means **read, write, and execute** permissions for the owner.
   - next three for group - `r-x` means **read and execute** permissions for the group.
   - next three for others- `r-x` means **read and execute** permissions for others.

   rwx (7)
   rw- (6)
   r-x (5)
   r-- (4)
   --x (1)

2. **Links**: (`1`)
   - The number of links or directories inside this directory. For a directory, this number (1) represents how many subdirectories, parent directories, or other references link to it. A hard link is like an alias or shortcut that points to the same file or directory in the filesystem. 

3. **Owner**: (`kavis`)
   - The username of the owner of the file or directory.

4. **Group**: (`197609`)
   - The group ID associated with the file.

5. **Size**: (`0`, `40`, `12288`, etc.)
   - The size of the file in bytes. For directories, it’s usually `0` since it doesn’t directly reflect the size of contents.

6. **Timestamp**: (`Oct 6 14:21`)
   - The date and time when the file or directory was last modified.

7. **File/Directory Name**: (`$Recycle.Bin`, `Documents and Settings`, etc.)
   - The name of the file or directory.

#### Listing Explanation:

**`Documents and Settings -> /c/Users`**:
   - This is a **symbolic link** (shortcut) that points to `/c/Users`. It means that trying to access `Documents and Settings` will actually redirect you to the `/c/Users` directory.

**`Drivers`**:
   - Likely a folder containing drivers for the system. Drivers are software that help the operating system communicate with hardware devices.

**`Program Files`** and **`Program Files (x86)`**:
   - Standard directories in Windows for installing 64-bit and 32-bit applications, respectively.

**`ProgramData`**:
   - A directory used to store application data that is not specific to a user. It’s shared among all users of the system.

**`Users`**:
    - This is where user profiles are stored on Windows, such as `Desktop`, `Documents`, `Downloads`, and so on for each user.

**`Windows`**:
    - The main directory containing the Windows operating system files.

**`hiberfil.sys`**:
    - A **system file** used by Windows to store the system state when it goes into **hibernation**. This file is often large and contains everything stored in memory (RAM).

In [95]:
%%bash
pwd
cd /
ls -im  #ls -i shows the inode number associated with each file or directory, which helps identify them uniquely
ls -l

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
12 bin, 8193 bin.usr-is-merged, 8194 boot, 1 dev, 16386 etc, 24577 home, 5 init,
1395 lib, 24578 lib.usr-is-merged, 1396 lib64, 11 lost+found, 32769 media,
32770 mnt, 40961 opt, 1 proc, 49153 root, 1 run, 1424 sbin,
57345 sbin.usr-is-merged, 57346 snap, 65537 srv, 1 sys, 73729 tmp, 73730 usr,
33554433 var
total 2284
lrwxrwxrwx   1 root root       7 Apr 22 13:08 bin -> usr/bin
drwxr-xr-x   2 root root    4096 Feb 26  2024 bin.usr-is-merged
drwxr-xr-x   2 root root    4096 Apr 22 13:08 boot
drwxr-xr-x  16 root root    3560 Oct  8 08:09 dev
drwxr-xr-x  92 root root    4096 Oct  8 08:09 etc
drwxr-xr-x   3 root root    4096 Oct  5 12:54 home
-rwxrwxrwx   1 root root 2260248 Sep 27 00:55 init
lrwxrwxrwx   1 root root       7 Apr 22 13:08 lib -> usr/lib
drwxr-xr-x   2 root root    4096 Apr  8  2024 lib.usr-is-merged
lrwxrwxrwx   1 root root       9 Apr 22 13:08 lib64 -> usr/lib64
drwx------   2 root root   16384 Oct  5 12:49 los

### inode and hardlink
The inode tracks the file or directory, while the hard link count tells you how many different references or paths point to that inode.

### inode
Each inode number (e.g., 12, 8193, 8194) uniquely identifies a file or directory on the filesystem, like a unique ID.

ls -i shows the inode number associated with each file or directory, which helps identify them uniquely


### hard link
Each directory starts with at least 2 hard links:

One for the directory itself (.)
One for its parent directory (..)
Every subdirectory inside it adds 1 more hard link to the count. This is because each subdirectory contains a link (..) pointing back to the parent directory. 

Example:
var has 13 hard links.
It includes . and ...
And 11 subdirectories inside var, each adding 1 to the count.

When you use ls -l, the number of hard links is shown in the second column. For example, in drwxr-xr-x 13 root root 4096 Oct 5 12:49 var, the 13 means there are 13 hard links to the var directory. The inode for var is linked to every file or directory that shares it. 

/bin -       essential command binaries - files in machine language for processor

/usr/bin -   User commands

/usr/lib -   Libraries

# Other commands

In [96]:
%%bash

date
date -R

cal

cal sep 1991

ncal sept 2024

Tue Oct  8 09:47:57 UTC 2024
Tue, 08 Oct 2024 09:47:57 +0000
    October 2024      
Su Mo Tu We Th Fr Sa  
       1  2  3  4  5  
 6  7  8  9 10 11 12  
13 14 15 16 17 18 19  
20 21 22 23 24 25 26  
27 28 29 30 31        
                      
   September 1991     
Su Mo Tu We Th Fr Sa  
 1  2  3  4  5  6  7  
 8  9 10 11 12 13 14  
15 16 17 18 19 20 21  
22 23 24 25 26 27 28  
29 30                 
                      
    September 2024    
Su  1  8 15 22 29   
Mo  2  9 16 23 30   
Tu  3 10 17 24      
We  4 11 18 25      
Th  5 12 19 26      
Fr  6 13 20 27      
Sa  7 14 21 28      


In [97]:
%%bash

free
free -h
groups

               total        used        free      shared  buff/cache   available
Mem:         7993724     1493904     6510220        3108      244560     6499820
Swap:        2097152           0     2097152
               total        used        free      shared  buff/cache   available
Mem:           7.6Gi       1.4Gi       6.2Gi       3.0Mi       238Mi       6.2Gi
Swap:          2.0Gi          0B       2.0Gi
kavis adm dialout cdrom floppy sudo audio dip video plugdev users netdev


In [104]:
%%bash
mkdir level1
ls -l
# permission string reflects permissions to users which can be changed my chmod command
chmod g-w level1
ls -l
rmdir level1


total 596
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 146714 Oct  8 09:42 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  8 09:50 level1
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
total 596
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 146714 Oct  8 09:42 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-r

### Difference Between Linux and Windows File Systems

### File System Types:

Linux: Uses file systems like ext4, ext3, xfs, and btrfs.
Windows: Uses NTFS (New Technology File System) and FAT32.


### Permissions
Linux: Has a detailed permission model with read, write, and execute permissions for the user, group, and others. It supports commands like chmod and chown to control access.
Windows: Uses Access Control Lists (ACLs) to set permissions. ACLs allow for more granular control but are not compatible with the Unix permission model. This is why changing permissions with chmod on a Windows file system has no effect.


### Case Sensitivity:

Linux: Case-sensitive. file.txt and File.txt are treated as separate files.
Windows: Generally case-insensitive. file.txt and File.txt are considered the same file.

### File Path Differences:

Linux: Paths use forward slashes (/), e.g., /home/kavis.
Windows: Paths use backslashes (\), e.g., C:\Users\kavis.

When you work in directories like /mnt/c/, /mnt/d/, etc., you’re actually interacting with the Windows file system through WSL, which has limitations when it comes to Unix-style operations.
By working in Linux directories like /home/kavis, you can use native Linux commands and features (like chmod) without restrictions.
In Summary: Move to directories like /home/kavis to use Linux file system features effectively and see your chmod changes reflect properly!

### Move to linux file system
To move to the Linux file system in WSL, you can navigate to the home directory or any other directory that resides within the Linux environment. Here's how you can do it:

Go to Your Linux Home Directory: cd~
Create a New Directory: Once you're in your Linux home directory, create a new directory for testing:


In [114]:
%%bash

cd ~
pwd

cd -    # change to previous directory
pwd

cd 
pwd

mkdir level1
ls -l

chmod g-x level1
ls -l

chmod g-r level1
ls -l

chmod o+w level1
ls -l

chmod 700 level1
ls -l

chmod 707 level1
ls -l

rm -r level1


/home/kavis
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/home/kavis
total 4
drwxr-xr-x 2 kavis kavis 4096 Oct  8 09:52 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwxr--r-x 2 kavis kavis 4096 Oct  8 09:52 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx---r-x 2 kavis kavis 4096 Oct  8 09:52 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx---rwx 2 kavis kavis 4096 Oct  8 09:52 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx------ 2 kavis kavis 4096 Oct  8 09:52 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx---rwx 2 kavis kavis 409

In [127]:
%%bash
cd 
pwd

mkdir level1
cd level1 
pwd
ls -l

touch file1 #make empty file - touch is actually used to chage timestamp
chmod 707 file1
pwd
ls -l

cp file1 file2
pwd
ls -l

mv file2 ..
pwd
ls -l

cd ..
pwd
ls -l

mv file2 file2a
ls -l

cd level1 
rm file1
cd ..
rm file2a
rm -r level1

/home/kavis
/home/kavis/level1
total 0
/home/kavis/level1
total 0
-rwx---rwx 1 kavis kavis 0 Oct  8 09:58 file1
/home/kavis/level1
total 0
-rwx---rwx 1 kavis kavis 0 Oct  8 09:58 file1
-rwx---r-x 1 kavis kavis 0 Oct  8 09:58 file2
/home/kavis/level1
total 0
-rwx---rwx 1 kavis kavis 0 Oct  8 09:58 file1
/home/kavis
total 4
-rwx---r-x 1 kavis kavis    0 Oct  8 09:58 file2
drwxr-xr-x 2 kavis kavis 4096 Oct  8 09:58 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
-rwx---r-x 1 kavis kavis    0 Oct  8 09:58 file2a
drwxr-xr-x 2 kavis kavis 4096 Oct  8 09:58 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1


In [128]:
%%bash
cd 
pwd
touch "file 2"  #space in filename
ls -li

alias rm="rm -i"
rm "file 2"
alias
pwd
ls -lia

mkdir level1a
ls -lia
# as new directory is added, hard links for present directory . increased by one

rm -r level1a

/home/kavis
total 0
83910 -rw-r--r-- 1 kavis kavis  0 Oct  8 09:59 file 2
43664 lrwxrwxrwx 1 kavis kavis 52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
alias rm='rm -i'
/home/kavis
total 52
  686 drwxr-x--- 9 kavis kavis 4096 Oct  8 09:59 .
24577 drwxr-xr-x 3 root  root  4096 Oct  5 12:54 ..
46407 -rw------- 1 kavis kavis 1911 Oct  8 02:58 .bash_history
46419 -rw-r--r-- 1 kavis kavis  220 Oct  5 12:54 .bash_logout
 1227 -rw-r--r-- 1 kavis kavis 3771 Oct  5 12:54 .bashrc
46417 drwx------ 3 kavis kavis 4096 Oct  5 13:50 .cache
82512 drwxr-xr-x 3 kavis kavis 4096 Oct  6 12:03 .dotnet
82473 drwxr-xr-x 3 kavis kavis 4096 Oct  6 12:06 .ipython
46428 drwxr-xr-x 2 kavis kavis 4096 Oct  6 10:49 .landscape
31011 drwxr-xr-x 3 kavis kavis 4096 Oct  6 11:12 .local
46423 -rw-r--r-- 1 kavis kavis    0 Oct  8 02:27 .motd_shown
46422 -rw-r--r-- 1 kavis kavis  807 Oct  5 12:54 .profile
46410 -rw-r--r-- 1 kavis kavis    0 Oct  6 10:52 .sudo_as_admin_successful
48219 drwxr-xr

In [130]:
%%bash
whoami
cd /   # move to root directory
pwd
ls -ma
# root folder doesnt have rwx permissions for kavis- it has permission only for root directory . 
touch file1
# write w not allowed in root directory . by other users 
clear

kavis
/
., .., bin, bin.usr-is-merged, boot, dev, etc, home, init, lib,
lib.usr-is-merged, lib64, lost+found, media, mnt, opt, proc, root, run, sbin,
sbin.usr-is-merged, snap, srv, sys, tmp, usr, var


touch: cannot touch 'file1': Permission denied


In [132]:
%%bash
cd /
cd var/log
pwd
head -n 2 alternatives.log  # Display the first 2 lines of the file

/var/log
update-alternatives 2024-09-27 20:13:01: run with --install /usr/bin/awk awk /usr/bin/mawk 5 --slave /usr/share/man/man1/awk.1.gz awk.1.gz /usr/share/man/man1/mawk.1.gz --slave /usr/bin/nawk nawk /usr/bin/mawk --slave /usr/share/man/man1/nawk.1.gz nawk.1.gz /usr/share/man/man1/mawk.1.gz
update-alternatives 2024-09-27 20:13:01: link group awk updated to point to /usr/bin/mawk


less alternatives.log      


How to read a full file ?

In [ ]:
%%bash

cd
pwd


# Create a test directory structure with nested subdirectories and files
mkdir -p test_dir/level1/level2
cd test_dir
touch file1 file2 level1/file3 level1/level2/file4
mkdir subdir1 subdir2

# 1. Default `ls` with no options
ls

# 2. Using `-l -t` vs `-t -l`:
#    `-l` is for long format and `-t` is for sorting by modification time.
#    These two options can be used in any order and the output will be the same.
ls -lt  # Long format sorted by modification time
ls -tl  # Sorted by modification time with long format

# 3. Using `-l` and `-S` in different orders:
#    `-S` sorts by size and `-l` shows long format. The order does not matter here.
ls -lS  # Long format sorted by file size
ls -Sl  # Sorted by file size in long format

# 4. Combining options `-l`, `-a`, and `-t` in different orders:
#    All options will work in any order, and the output will be the same.
ls -lat  # Long format, including hidden files, sorted by modification time
ls -tal  # Sorted by modification time, long format, including hidden files
ls -alt  # Including hidden files, long format, sorted by modification time

# 5. Using `-d` with directory arguments:
#    `-d` changes how directories are listed. The order of `-d` matters if you include it with other options.
#    When `-d` is specified after other options, it will change how the directories themselves are displayed.
ls -ld level1 subdir1  # Show details of directories level1 and subdir1, not their contents
ls -dl level1 subdir1  # Same output as above since -d modifies the behavior of -l

# 6. Recursive listing (`-R`) combined with other options in different orders:
#    `-R` will recursively list all contents. Combining with other options like `-l`, `-a`, or `-S` can be in any order.
ls -lRa  # Long format, recursive, including hidden files
ls -aRl  # Including hidden files, recursive, long format

# Clean up: remove the created test directory
cd ..
rm -r test_dir


Key Takeaways on Option Ordering:
Most options can be combined in any order:

For example, ls -lS, ls -Sl, and ls -l -S will produce the same output because -l and -S do not conflict.
Some options modify the behavior of other options:

ls -d (show directory itself) when combined with -l (ls -ld) will show details of the directory itself rather than its contents. But using -l -d or -d -l does not change the behavior.
Grouping of options:

Options can be grouped together as ls -l -t or ls -lt. There is no difference in output when options are grouped or not, as long as the options do not modify each other’s behavior.
Effect of -R (recursive):

The order of -R does not matter when combined with other options like -l or -a. All contents will be listed recursively no matter where -R appears.
Options that change display behavior (e.g., -d, -l):

These options need to be understood carefully, as -d will override the behavior of other options like -R (no recursion).
Summary
The order of options does not matter in most cases, but options like -d can change how others behave, so be mindful when combining.

In [15]:
%%bash

cd 
pwd

# Create a test directory structure for demonstration
mkdir -p test_recursive/dir1/dir2
cd test_recursive
touch file1 file2 dir1/file3 dir1/dir2/file4
mkdir dir3

# 1. Basic Recursive Listing (-R)
#    Lists the contents of the current directory and all its subdirectories.
ls -R

# 2. Recursive Listing in Long Format (-lR)
#    Long format listing, showing permissions, owners, sizes, timestamps, and file names recursively.
ls -lR

# 3. Recursive Listing with Hidden Files (-aR)
#    Includes hidden files (those starting with a dot `.`) in the listing.
ls -aR

# 4. Recursive Listing with Long Format and Hidden Files (-laR)
#    Combines long format and hidden files with recursive listing.
ls -laR

# 5. Recursive Listing for a Specific Directory (dir1)
#    Lists all contents of `dir1` and its subdirectories, recursively.
ls -R dir1

# 6. Recursive Listing for Multiple Directories (dir1, dir3)
#    Lists all contents of `dir1` and `dir3` and their subdirectories recursively.
ls -R dir1 dir3

# 7. Recursive Listing with Human-Readable Sizes (-lhR)
#    Shows sizes in KB, MB, etc., for better readability.
ls -lhR

# 8. Recursive Listing with Sorting by Modification Time (-ltR)
#    Recursively lists all contents, sorting each directory’s contents by modification time.
ls -ltR

# 9. Recursive Listing in Reverse Order (-lrR)
#    Recursively lists all contents, displaying each directory’s contents in reverse order.
ls -lrR

# 10. Recursive Listing and Limiting Output to Directories (-dR)
#     Show information about directories themselves, without listing their contents.
#     NOTE: -d overrides the recursive behavior, so only directories are listed.
ls -dR */

# 11. Combining -R with -s to Show File Sizes in Blocks
#     Displays file sizes in blocks recursively through the directory tree.
ls -sR

# Clean up: remove the created test directory
cd ..
rm -r test_recursive


/home/kavis
.:
dir1
dir3
file1
file2

./dir1:
dir2
file3

./dir1/dir2:
file4

./dir3:
.:
total 8
drwxr-xr-x 3 kavis kavis 4096 Oct  8 02:48 dir1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:48 dir3
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:48 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:48 file2

./dir1:
total 4
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:48 dir2
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:48 file3

./dir1/dir2:
total 0
-rw-r--r-- 1 kavis kavis 0 Oct  8 02:48 file4

./dir3:
total 0
.:
.
..
dir1
dir3
file1
file2

./dir1:
.
..
dir2
file3

./dir1/dir2:
.
..
file4

./dir3:
.
..
.:
total 16
drwxr-xr-x  4 kavis kavis 4096 Oct  8 02:48 .
drwxr-x--- 11 kavis kavis 4096 Oct  8 02:48 ..
drwxr-xr-x  3 kavis kavis 4096 Oct  8 02:48 dir1
drwxr-xr-x  2 kavis kavis 4096 Oct  8 02:48 dir3
-rw-r--r--  1 kavis kavis    0 Oct  8 02:48 file1
-rw-r--r--  1 kavis kavis    0 Oct  8 02:48 file2

./dir1:
total 12
drwxr-xr-x 3 kavis kavis 4096 Oct  8 02:48 .
drwxr-xr-x 4 kavis kavis 4096 Oct  8 02:48 ..
drwx

Explanation of Recursive Listing Options:


ls -R: This is the basic recursive listing. It displays all files and directories under the current directory and its subdirectories. It’s useful for getting a complete overview of the entire directory tree.


ls -lR: Combines the long format listing (-l) with recursive listing. This provides a detailed view of permissions, owners, sizes, and modification timestamps for all files and directories recursively.


ls -aR: Shows all files, including hidden ones (files starting with .), recursively.


ls -laR: Combines long format and hidden file listing with recursive listing.


ls -R <directory_name>: Lists the contents of a specific directory (e.g., dir1) recursively. It will show all files and subdirectories under dir1.


ls -lhR: Combines long format with human-readable file sizes (KB, MB, etc.) for all files and directories recursively.


ls -ltR: Combines sorting by modification time (-t) with recursive listing. This will show the most recently modified files first within each directory.


ls -lrR: Combines reverse sorting (-r) with recursive listing. This will display files in reverse alphabetical order within each directory.


ls -dR: Using -d with -R will only display the directory names themselves, not their contents. This option can be useful if you want to list the directories without listing everything inside.


ls -sR: Shows file sizes in blocks, recursively. This is useful for getting an idea of storage consumption in each directory.


When to Use Recursive Listing:
To see the structure of nested directories.
To analyze large directory trees for storage usage or permissions.
To find specific files deep within a directory structure without manually traversing each directory.
Using recursive listing helps in scenarios where you want to understand the overall structure and contents of a directory tree.

In [1]:
%%bash
cd /
cd var/log
pwd 
chmod o+w alternatives.log  # permissions change not allowed by users other than root directory 
mkdir level1                # write not allowed by users other than root directory 
touch file1
clear

/var/log


chmod: changing permissions of 'alternatives.log': Operation not permitted
mkdir: cannot create directory ‘level1’: Permission denied
touch: cannot touch 'file1': Permission denied


In [18]:
%%bash
pwd
cd 
pwd
cd rep1/Linux_shell
file Linux_shell_w1.ipynb   #what kind of file is it
file 1500x500.jpeg          #its a binary file, not a text file
file random_toss.py
file level1

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/home/kavis
Linux_shell_w1.ipynb: Unicode text, UTF-8 text, with very long lines (1311)
1500x500.jpeg: JPEG image data, JFIF standard 1.01, aspect ratio, density 1x1, segment length 16, progressive, precision 8, 1500x500, components 3
random_toss.py: ASCII text, with CRLF line terminators
level1: directory


In [22]:
%%bash
date

cd /rep1/Linux_shell
ls -l

touch SC_Week0_Unit0.ipynb   #changes the timestamp to latest time
ls -l


Sun Oct  6 19:54:49 UTC 2024


bash: line 3: cd: /rep1/Linux_shell: No such file or directory


total 544
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  90847 Oct  6 19:32 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  6 18:19 level1
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
total 544
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  90847 Oct  6 19:32 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-r

In [10]:

%%bash 

cd 
pwd

cd ..
pwd

cd ..
pwd

cd ..
pwd

cd //home///kavis
pwd


/home/kavis
/home
/
/
//home/kavis


In [151]:
%%bash
# Create a sample directory structure with files for demonstration
cd
mkdir -p text_demo
cd text_demo

# Create sample text files
echo -e "Hello World\nThis is a sample text file." > file1.txt
echo -e "Line 1\nLine 2\nLine 3\nLine 4\nLine 5" > file2.txt
echo -e "First Line\nSecond Line\nThird Line" > file3.txt
echo -e "Line A\nLine B\nLine C\nLine D" > file4.txt

# List the files created to verify
ls -l

# Display contents of the files using `cat`
cat file1.txt          # Display contents of file1.txt
cat file2.txt file3.txt  # Concatenate and display contents of multiple files

# Display contents of a file page by page using `more`
more file2.txt         # Display file2.txt one page at a time

# Display contents of a file using `less` (allows scrolling up and down)
less file3.txt         # View file3.txt with less (q to quit)

# Count the number of lines, words, and characters in a file using `wc`
wc file1.txt           # Show line, word, and character count of file1.txt
wc -l file2.txt        # Show only the number of lines in file2.txt
wc -w file3.txt        # Show only the number of words in file3.txt
wc -c file4.txt        # Show only the number of characters in file4.txt
wc -m file1.txt        # Show the number of characters including multi-byte characters

# Display the first few lines of a file using `head`
head file1.txt         # Show first 10 lines of file1.txt (default)
head -n 2 file2.txt    # Show first 2 lines of file2.txt
head -n 5 file3.txt    # Show first 5 lines of file3.txt

# Display the last few lines of a file using `tail`
tail file1.txt         # Show last 10 lines of file1.txt (default)
tail -n 2 file2.txt    # Show last 2 lines of file2.txt
tail -n 5 file3.txt    # Show last 5 lines of file3.txt

# Combine commands to extract specific lines using `head` and `tail`
head -n 4 file2.txt | tail -n 2  # Show lines 3 and 4 of file2.txt

# Clean up: remove the created text directory
cd ..
rm -r text_demo


total 16
-rw-r--r-- 1 kavis kavis 40 Oct  8 10:28 file1.txt
-rw-r--r-- 1 kavis kavis 35 Oct  8 10:28 file2.txt
-rw-r--r-- 1 kavis kavis 34 Oct  8 10:28 file3.txt
-rw-r--r-- 1 kavis kavis 28 Oct  8 10:28 file4.txt
Hello World
This is a sample text file.
Line 1
Line 2
Line 3
Line 4
Line 5
First Line
Second Line
Third Line

# Display contents of a file using `less` (allows scrolling up and down)
less file3.txt         # View file3.txt with less (q to quit)

# Count the number of lines, words, and characters in a file using `wc`
wc file1.txt           # Show line, word, and character count of file1.txt
wc -l file2.txt        # Show only the number of lines in file2.txt
wc -w file3.txt        # Show only the number of words in file3.txt
wc -c file4.txt        # Show only the number of characters in file4.txt
wc -m file1.txt        # Show the number of characters including multi-byte characters

# Display the first few lines of a file using `head`
head file1.txt         # Show first 10 lines

Explanation of Each Command
cat:

cat file1.txt: Displays the contents of file1.txt.
cat file2.txt file3.txt: Concatenates and displays the contents of file2.txt and file3.txt.
more:

more file2.txt: Displays file2.txt one page at a time. You can scroll down by pressing Enter or Space.
less:

less file3.txt: Displays file3.txt in a scrollable view. You can scroll up and down using arrow keys, and quit by pressing q.
wc:

wc file1.txt: Shows the number of lines, words, and characters in file1.txt.
wc -l file2.txt: Shows only the number of lines in file2.txt.
wc -w file3.txt: Shows only the number of words in file3.txt.
wc -c file4.txt: Shows only the number of characters in file4.txt.
wc -m file1.txt: Shows the number of characters, including multi-byte characters (like UTF-8).
head:

head file1.txt: Shows the first 10 lines of file1.txt (default behavior).
head -n 2 file2.txt: Shows the first 2 lines of file2.txt.
head -n 5 file3.txt: Shows the first 5 lines of file3.txt.
tail:

tail file1.txt: Shows the last 10 lines of file1.txt (default behavior).
tail -n 2 file2.txt: Shows the last 2 lines of file2.txt.
tail -n 5 file3.txt: Shows the last 5 lines of file3.txt.
Combining head and tail:

head -n 4 file2.txt | tail -n 2: Extracts and shows lines 3 and 4 of file2.txt by first selecting the first 4 lines with head and then selecting the last 2 of those with tail.

In [150]:
%%bash
# Create a sample directory structure
cd
mkdir -p command_demo
cd command_demo

# Create sample script files to use with the `which` and `type` commands
echo -e "#!/bin/bash\necho 'This is a sample script'" > sample_script.sh
chmod +x sample_script.sh

# 1. `man` command
# Display the manual page for the `ls` command
#man ls

# Display the manual page for the `man` command itself
#man man

# 2. `which` command
# Locate the executable path for commonly used commands
which ls            # Locate the `ls` command
which bash          # Locate the `bash` command
which sample_script.sh  # Locate the `sample_script.sh` script

# 3. `apropos` command
# Search the manual pages for a keyword, like 'list' to find related commands
#apropos list        # Search for commands related to 'list'
#apropos copy        # Search for commands related to 'copy'

# 4. `info` command
# Display the info documentation for specific commands
#info ls             # Display the `info` page for the `ls` command
#info bash           # Display the `info` page for the `bash` shell

# 5. `whatis` command
# Get a short description of a command
whatis ls           # Display a short description of `ls`
whatis bash         # Display a short description of `bash`

# 6. `help` command (for shell built-ins)
# Display help information for shell built-in commands
#help cd             # Display help for the built-in command `cd`
#help alias          # Display help for the built-in command `alias`
#help type           # Display help for the built-in command `type`

# 7. `type` command
# Show how a command will be interpreted (built-in, alias, function, file)
#type ls             # Show how `ls` is interpreted (usually an executable)
#type cd             # Show how `cd` is interpreted (usually a shell built-in)
#type sample_script.sh  # Show how `sample_script.sh` is interpreted (file path)

# Clean up: remove the created command_demo directory
cd ..
rm -r command_demo


/bin/ls
/bin/bash
ls... (1)            - list directory contents
bash... (1)          - GNU Bourne-Again SHell


Explanation of Each Command
man (Manual Pages):

man ls: Displays the manual page for the ls command, including detailed descriptions, options, and usage examples.
man man: Displays the manual page for the man command itself, explaining how to use it.
which (Locate Executables):

which ls: Shows the full path of the ls executable.
which bash: Shows the full path of the bash shell executable.
which sample_script.sh: Shows the location of the sample_script.sh script file in your PATH if it's executable and found.
apropos (Search Manual Pages):

apropos list: Searches for commands related to "list". It shows a list of commands and their short descriptions that match the keyword "list".
apropos copy: Searches for commands related to "copy". Useful when you know what a command should do but not the exact command name.
info (GNU Info Pages):

info ls: Displays the info documentation for the ls command, which often includes more detailed information than man.
info bash: Displays the info documentation for the bash shell.
whatis (Short Descriptions):

whatis ls: Displays a brief description of the ls command.
whatis bash: Displays a brief description of the bash shell.
help (Shell Built-ins Help):

help cd: Displays help information for the cd built-in command.
help alias: Displays help information for the alias built-in command.
help type: Displays help information for the type built-in command.
type (Describe a Command Type):

type ls: Shows that ls is an executable file.
type cd: Shows that cd is a shell built-in command.
type sample_script.sh: Shows how sample_script.sh is interpreted (e.g., as a file in the current directory).

In [146]:
%%bash
# Create a sample directory structure for demonstration
cd
mkdir -p arg_demo/subdir1 arg_demo/subdir2
cd arg_demo

# Create sample files for testing
echo "This is file1" > file1.txt
echo "This is file2" > file2.txt
echo "This is file3" > file3.txt

# 1. Multiple Arguments
# Using `mv` and `cp` commands with multiple source files as arguments
ls                # List files to see what's present

# Move multiple files (file1.txt and file2.txt) to subdir1
mv file1.txt file2.txt subdir1/
ls subdir1        # Check that both files were moved to subdir1

# Copy multiple files (file1.txt and file2.txt) back to the main directory
cp subdir1/file1.txt subdir1/file2.txt .
ls                # Check that both files were copied back to the main directory

# 2. Second Argument
# The second argument is typically the target location for `mv` and `cp` commands.
# Move a file to subdir2, where subdir2 is the second argument.
mv file3.txt subdir2/
ls subdir2        # Check that file3.txt was moved to subdir2

# Copy the file back to the main directory, using subdir2/file3.txt as source and . as the second argument.
cp subdir2/file3.txt .
ls                # Check that file3.txt was copied back to the main directory

# 3. Interpretation of Last Argument
# The last argument is interpreted as the target directory if multiple sources are provided.
# Move file1.txt and file2.txt into subdir2 using subdir2/ as the last argument.
mv file1.txt file2.txt subdir2/
ls subdir2        # Check that both files are moved to subdir2

# Copy file1.txt and file2.txt back to the main directory using . as the last argument.
cp subdir2/file1.txt subdir2/file2.txt .
ls                # Check that both files are copied back to the main directory

# 4. Recursion Assumed for `mv` but Not for `cp`
# Create nested directories and files for demonstration of recursion
mkdir -p subdir1/nested_dir
echo "Nested file content" > subdir1/nested_dir/nested_file.txt

# Moving a directory with contents is straightforward with `mv`:
mv subdir1/nested_dir subdir2/
ls subdir2/nested_dir  # Check that nested_dir and its contents were moved to subdir2

# Copying a directory requires the `-r` or `-R` flag for recursion:
# Copy nested_dir from subdir2 back to subdir1 using `-r` for recursion.
cp -r subdir2/nested_dir subdir1/
ls subdir1/nested_dir  # Check that nested_dir and its contents were copied back to subdir1

# Clean up: remove the created arg_demo directory
cd ..
rm -r arg_demo


file1.txt
file2.txt
file3.txt
subdir1
subdir2
file1.txt
file2.txt
file1.txt
file2.txt
file3.txt
subdir1
subdir2
file3.txt
file1.txt
file2.txt
file3.txt
subdir1
subdir2
file1.txt
file2.txt
file3.txt
file1.txt
file2.txt
file3.txt
subdir1
subdir2
nested_file.txt
nested_file.txt


Explanation of Each Section:
Multiple Arguments:

The mv and cp commands can take multiple source files as arguments followed by a destination directory.
Example: mv file1.txt file2.txt subdir1/ moves both file1.txt and file2.txt to subdir1.
Second Argument:

The second argument is typically the destination or target for the operation.
Example: mv file3.txt subdir2/ moves file3.txt to the subdir2 directory.
Interpretation of Last Argument:

The last argument is always interpreted as the target if multiple sources are provided.
Example: mv file1.txt file2.txt subdir2/ moves file1.txt and file2.txt into subdir2 because the last argument (subdir2/) is considered the destination.
Recursion Assumed for mv but Not for cp:

mv assumes recursion when moving directories with contents. You can move a directory without specifying additional flags.
Example: mv subdir1/nested_dir subdir2/ moves the entire nested_dir with its contents to subdir2.
cp requires -r or -R flag for recursion. Without the -r flag, copying a directory will result in an error.
Example: cp -r subdir2/nested_dir subdir1/ copies the entire nested_dir with its contents back to subdir1.
Summary of Commands Used:
mv:

Moves files or directories from one location to another.
Recursively moves directories by default.
cp:

Copies files from one location to another.
Requires -r or -R for copying directories and their contents.
ls:

Lists the contents of directories to verify file operations.

In [145]:
%%bash
# Create a sample directory structure for demonstration
cd
mkdir -p link_demo
cd link_demo

# Create a sample file for linking demonstration
echo "This is the original file." > original_file.txt
echo "This is another file." > another_file.txt

# 1. **Hard Links**:
# Create a hard link to the original file
# Syntax: ln <source_file> <hard_link_name>
ln original_file.txt hard_link.txt

# List files to see if hard_link.txt is created
ls -li            # Use -i to show inode numbers

# Display contents of both files to verify they are identical
cat original_file.txt
cat hard_link.txt

# Make a change to the hard link and verify if it affects the original file
echo "Appending text to hard link." >> hard_link.txt

# Display contents again to verify changes reflect in both files
cat original_file.txt
cat hard_link.txt

# Check inode numbers of both files to confirm they are the same
ls -li original_file.txt hard_link.txt

# 2. **Symbolic Links (Soft Links)**:
# Create a symbolic link to the original file
# Syntax: ln -s <source_file> <symbolic_link_name>
ln -s original_file.txt soft_link.txt

# List files to see if soft_link.txt is created
ls -l            # Use -l to show symbolic links with the arrow (->)

# Display contents of the symbolic link to verify it points to the original file
cat soft_link.txt

# Make a change to the original file and see if it reflects in the symbolic link
echo "Appending text to original file." >> original_file.txt

# Display contents again to verify changes reflect in the symbolic link
cat soft_link.txt

# Check inode numbers of all three files to confirm hard link shares inode but symbolic link does not
ls -li original_file.txt hard_link.txt soft_link.txt

# 3. **Delete the Original File**:
# Delete the original file and see what happens to the hard and symbolic links
rm original_file.txt

# List files to see the result
ls -l

# Display contents of the remaining files
cat hard_link.txt      # Should work as it shares the same inode as the original file
cat soft_link.txt      # Will show an error since the original file is deleted

# Clean up: remove the created link_demo directory
cd ..
rm -r link_demo


total 12
83912 -rw-r--r-- 1 kavis kavis 22 Oct  8 10:25 another_file.txt
83911 -rw-r--r-- 2 kavis kavis 27 Oct  8 10:25 hard_link.txt
83911 -rw-r--r-- 2 kavis kavis 27 Oct  8 10:25 original_file.txt
This is the original file.
This is the original file.
This is the original file.
Appending text to hard link.
This is the original file.
Appending text to hard link.
83911 -rw-r--r-- 2 kavis kavis 56 Oct  8 10:25 hard_link.txt
83911 -rw-r--r-- 2 kavis kavis 56 Oct  8 10:25 original_file.txt
total 12
-rw-r--r-- 1 kavis kavis 22 Oct  8 10:25 another_file.txt
-rw-r--r-- 2 kavis kavis 56 Oct  8 10:25 hard_link.txt
-rw-r--r-- 2 kavis kavis 56 Oct  8 10:25 original_file.txt
lrwxrwxrwx 1 kavis kavis 17 Oct  8 10:25 soft_link.txt -> original_file.txt
This is the original file.
Appending text to hard link.
This is the original file.
Appending text to hard link.
Appending text to original file.
83911 -rw-r--r-- 2 kavis kavis 89 Oct  8 10:25 hard_link.txt
83911 -rw-r--r-- 2 kavis kavis 89 Oct  8 10:25

cat: soft_link.txt: No such file or directory


Explanation of Each Section:
Creating Hard Links:

Command: ln original_file.txt hard_link.txt
A hard link is another name for the same file. Both the original file and the hard link share the same inode number.
Changes to the hard link reflect in the original file and vice versa.
If you delete the original file, the hard link still exists and retains the file content.
Creating Symbolic (Soft) Links:

Command: ln -s original_file.txt soft_link.txt
A symbolic (soft) link is a pointer to the original file. It does not share the same inode number as the original file.
If you delete the original file, the symbolic link becomes a "broken link" and no longer works.
Symbolic links are displayed with an arrow (->) pointing to the target file in the output of ls -l.
Verifying Links:

Command: ls -li and cat <filename>
The -i option shows inode numbers, allowing you to verify that hard links share the same inode number while symbolic links do not.
Displaying file contents (cat) demonstrates that changes made to the hard link reflect in the original file, while the symbolic link depends on the presence of the original file.
Effect of Deleting the Original File:

After deleting the original file (rm original_file.txt):
Hard Link: The hard link continues to exist because it shares the inode with the original file.
Symbolic Link: The symbolic link becomes broken because it points to a deleted file.
Summary of Commands Used:
Creating Hard Links:

ln <source_file> <hard_link_name>: Creates a hard link named <hard_link_name> pointing to <source_file>.
Creating Symbolic Links:

ln -s <source_file> <symbolic_link_name>: Creates a symbolic link named <symbolic_link_name> pointing to <source_file>.
Verifying Links:

ls -li: Shows files with their inode numbers to confirm hard links share the same inode.
cat <file_name>: Displays file contents to show the relationship between links and the original file.
Deleting Original File:

rm <file_name>: Deletes the original file to see the behavior of hard and symbolic links.

In [143]:
%%bash
# Create a sample directory structure for demonstration
cd
mkdir -p size_demo
cd size_demo

# Create sample files of varying sizes
echo "This is a small file." > small_file.txt
dd if=/dev/zero of=medium_file.txt bs=512 count=100  # Creates a 50 KB file
dd if=/dev/zero of=large_file.txt bs=1024 count=200  # Creates a 200 KB file

# List the created files
ls -lh            # List files with human-readable sizes

# 1. **File Size Using `ls -s`**:
# Display file sizes in blocks with `ls -s`
ls -s             # Shows the size of each file in 1KB blocks (default)

# Use `-h` to display file sizes in human-readable format along with `-s`
ls -sh            # Shows sizes in human-readable format

# List with detailed view including file sizes and other attributes
ls -lsh           # Combine -l, -s, and -h to show detailed information with file sizes

# 2. **File Size Using `stat`**:
# Display detailed information including file size and block size
stat small_file.txt   # Shows file size in bytes, allocated blocks, and block size
stat medium_file.txt  # Shows file size in bytes, allocated blocks, and block size
stat large_file.txt   # Shows file size in bytes, allocated blocks, and block size

# 3. **File Size Using `du` (Disk Usage)**:
# Display disk usage of each file in blocks
du small_file.txt     # Shows the disk usage of small_file.txt
du medium_file.txt    # Shows the disk usage of medium_file.txt
du large_file.txt     # Shows the disk usage of large_file.txt

# Show disk usage in human-readable format
du -h small_file.txt
du -h medium_file.txt
du -h large_file.txt

# Show total size of the current directory and its contents
du -sh .             # Total disk usage of the current directory

# 4. **Role of Block Size**:
# Create a new file with a specific block size using `dd`
dd if=/dev/zero of=block_size_file.txt bs=4096 count=1  # Creates a file with 1 block of 4KB

# Display block size using `stat` to see allocation
stat block_size_file.txt

# Compare `stat` and `du` output for the file with a single block
stat block_size_file.txt
du block_size_file.txt

# Check file size in blocks using `ls -s`
ls -s block_size_file.txt

# Clean up: remove the created size_demo directory
cd ..
rm -r size_demo


100+0 records in
100+0 records out
51200 bytes (51 kB, 50 KiB) copied, 0.000129264 s, 396 MB/s
200+0 records in
200+0 records out
204800 bytes (205 kB, 200 KiB) copied, 0.000378211 s, 541 MB/s


total 256K
-rw-r--r-- 1 kavis kavis 200K Oct  8 10:23 large_file.txt
-rw-r--r-- 1 kavis kavis  50K Oct  8 10:23 medium_file.txt
-rw-r--r-- 1 kavis kavis   22 Oct  8 10:23 small_file.txt
total 256
200 large_file.txt
 52 medium_file.txt
  4 small_file.txt
total 256K
200K large_file.txt
 52K medium_file.txt
4.0K small_file.txt
total 256K
200K -rw-r--r-- 1 kavis kavis 200K Oct  8 10:23 large_file.txt
 52K -rw-r--r-- 1 kavis kavis  50K Oct  8 10:23 medium_file.txt
4.0K -rw-r--r-- 1 kavis kavis   22 Oct  8 10:23 small_file.txt
  File: small_file.txt
  Size: 22        	Blocks: 8          IO Block: 4096   regular file
Device: 8,32	Inode: 83911       Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   kavis)   Gid: ( 1000/   kavis)
Access: 2024-10-08 10:23:48.784920792 +0000
Modify: 2024-10-08 10:23:48.784920792 +0000
Change: 2024-10-08 10:23:48.784920792 +0000
 Birth: 2024-10-08 10:23:48.784920792 +0000
  File: medium_file.txt
  Size: 51200     	Blocks: 104        IO Block: 4096   regular file


1+0 records in
1+0 records out
4096 bytes (4.1 kB, 4.0 KiB) copied, 2.0926e-05 s, 196 MB/s


Explanation of Each Command:
File Size Using ls -s:

ls -s shows the size of each file in blocks. By default, each block is 1KB on most systems.
ls -sh displays file sizes in human-readable format (KB, MB, GB).
ls -lsh combines -l, -s, and -h to show detailed information along with file sizes in blocks and human-readable format.
File Size Using stat:

stat provides detailed information about a file, including:
File size: Actual file size in bytes.
Blocks allocated: Number of blocks allocated to store the file on disk.
IO Block size: The block size used for file system I/O operations.
This command helps you understand the difference between the file size and the space occupied due to block allocation.
File Size Using du (Disk Usage):

du shows the actual disk usage of a file, which may be larger than the file's size due to the file system's block size.
du -h <filename> displays the disk usage in a human-readable format.
du -sh . shows the total disk usage of the current directory and its contents.
Role of Block Size:

Creating a file with a specific block size using dd helps illustrate how block size affects disk usage.
stat displays the block size used for I/O operations and the number of blocks allocated.
du shows the actual disk usage, which will reflect the allocated blocks.
Key Concepts:
File Size vs. Disk Usage:

The actual file size is the number of bytes in the file.
Disk usage is the amount of disk space the file occupies, which depends on the file system's block size.
Block Size:

Files are stored in blocks on disk. If a file is smaller than the block size, it still occupies one full block.
For example, a 1-byte file on a system with a 4KB block size will occupy 4KB of disk space.
Understanding Block Size with stat and du:

stat shows file size, blocks allocated, and I/O block size.
du shows the actual disk usage in terms of blocks allocated.

In [144]:
%%bash
# Exploring the /proc filesystem

# 1. List contents of the /proc directory to see various files and subdirectories
ls -l /proc

# 2. Display the contents of some common files in /proc
# These files provide system information, process status, and other kernel parameters.

# Display system information from the /proc/cpuinfo file
cat /proc/cpuinfo | head -n 20    # Show the first 20 lines of CPU information

# Display memory information from the /proc/meminfo file
cat /proc/meminfo | head -n 20    # Show the first 20 lines of memory information

# Show the current kernel version from the /proc/version file
cat /proc/version

# Display uptime information from the /proc/uptime file
cat /proc/uptime

# Display current processes and their status
# The /proc directory contains subdirectories named after each process ID (PID).
ls -l /proc | grep "^[0-9]"        # List only the directories representing PIDs

# Display detailed information of a specific process (e.g., process ID 1)
cat /proc/1/status                 # Show the status of process with PID 1

# Display the command line of a specific process (e.g., process ID 1)
cat /proc/1/cmdline

# 3. Inspecting files in /proc for additional system details
# Display partitions information from /proc/partitions
cat /proc/partitions

# Display current mounts from /proc/mounts
cat /proc/mounts

# 4. Exploring the /sys filesystem
# /sys is used to expose kernel modules, devices, and other hardware information.

# List contents of the /sys directory
ls -l /sys

# List contents of /sys/class to see different device classes
ls -l /sys/class

# Display information about a specific device (e.g., CPU)
ls -l /sys/devices/system/cpu

# Show all CPUs present on the system
ls -l /sys/devices/system/cpu/cpu*

# Display information about a specific CPU (e.g., CPU0)
cat /sys/devices/system/cpu/cpu0/topology/core_id
cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor

# 5. Check block device information using /sys/block
ls -l /sys/block

# Show details about a specific block device (e.g., sda)
ls -l /sys/block/sda

# Clean up: No cleanup required since /proc and /sys are in-memory filesystems


total 0
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 1
dr-xr-xr-x  9 systemd-resolve  systemd-resolve                0 Oct  8 08:09 115
dr-xr-xr-x  9 systemd-timesync systemd-timesync               0 Oct  8 08:09 121
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 149
dr-xr-xr-x  9 messagebus       messagebus                     0 Oct  8 08:09 150
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 157
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 160
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 165
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 175
dr-xr-xr-x  9 syslog           syslog                         0 Oct  8 08:09 185
dr-xr-xr-x  9 root             root                           0 Oct  8 08:09 192
dr-xr-xr-x  9 kavis            kavis                          0 Oct  8 08:15 1920
dr-xr-xr-x  9 root   

cat: /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor: No such file or directory


Explanation of Each Command:
Exploring /proc Filesystem:

The /proc directory is a virtual filesystem that provides information about the kernel, system, and running processes. The contents are not actual files but rather kernel-generated information.
The following files and directories are explored:
/proc/cpuinfo: Shows detailed information about the CPUs on the system.
/proc/meminfo: Shows memory-related information such as total memory, free memory, and cached memory.
/proc/version: Displays the Linux kernel version.
/proc/uptime: Shows system uptime in seconds.
/proc/[PID]: Each subdirectory corresponds to a running process identified by its Process ID (PID).
/proc/partitions: Displays partition information.
/proc/mounts: Shows all mounted filesystems and their properties.
Exploring /sys Filesystem:

The /sys directory is another virtual filesystem used to provide information about devices, kernel modules, and hardware attributes.
Common subdirectories in /sys:
/sys/class: Represents different classes of devices, such as cpu, net, and block.
/sys/devices: Provides detailed information about devices connected to the system.
/sys/block: Represents block devices (e.g., storage devices).
Checking Hardware and Device Information:

/sys/devices/system/cpu/cpu*: Lists all CPUs present on the system.
/sys/block/sda: Represents a block device such as a hard disk or SSD.
Detailed Information on Processes:

Each process running on the system has a corresponding directory in /proc with its PID as the directory name.
status, cmdline, stat, and other files within these directories provide information about the process, such as its command-line arguments, memory usage, and more.
Key Concepts:
In-memory Filesystems:

/proc and /sys are virtual filesystems that exist in memory. They do not occupy disk space and are used by the kernel to expose information and configuration options.
Role of /proc:

/proc primarily provides information about the kernel, system resources, and running processes. It is read-only, and changing files here will not persist across reboots.
Role of /sys:

/sys provides a structured view of kernel modules, devices, and hardware attributes. It allows interaction with hardware settings (e.g., changing CPU frequency).
